tests/test_assignment.py

```diff
@@ -76,3 +88,80 @@ def _prepare_submission(self):
 
         self.assertEqual(total_score, 0)
         self.assertEqual(max_score, 1)
+
+    def test_hard_fail(self):
+        test_cases = [
+            (
+                [TestAssignment.QuestionBase(1), TestAssignment.QuestionBase(1)],
+                [(False, False), (False, False)]
+            ),
+            (
+                [TestAssignment.QuestionBase(1), TestAssignment.QuestionHardFail(1)],
+                [(False, False), (True, False)]
+            ),
+            (
+                [TestAssignment.QuestionHardFail(1), TestAssignment.QuestionBase(1)],
+                [(True, False), (False, True)]
+            ),
+            (
+                [TestAssignment.QuestionHardFail(1), TestAssignment.QuestionHardFail(1)],
+                [(True, False), (False, True)]
+            ),
+        ]
+
+        class TA(autograder.assignment.Assignment):
+            def _prepare_submission(self):
+                return lambda: False
+
+        for i in range(len(test_cases)):
+            with self.subTest(i = i):
+                questions, expected_results = test_cases[i]
+
+                assignment = TA(f'test_hard_fail_{i}', questions)
+                result = assignment.grade(show_exceptions = True)
+
+                total_score, max_score = result.get_score()
+
+                self.assertEqual(total_score, 0)
+                self.assertEqual(max_score, 2)
+
+                for j in range(len(expected_results)):
+                    (expected_hard_fail, expected_skipped) = expected_results[j]
+
+                    self.assertEqual(expected_hard_fail, result.questions[j].hard_fail)
+                    self.assertEqual(expected_skipped, result.questions[j].skipped)
+
+                    if (expected_skipped):
+                        self.assertIn(SKIPPING_QUESTION_MESSAGE, result.questions[j].message)
+                    else:
+                        self.assertIn(BASE_ERROR_MESSAGE, result.questions[j].message)
+
+    def test_hard_fail_full_credit(self):
```

Comment:
Both of your tests feel a bit incomplete.
You never hit a case where a submission gets 1 out of 2.
Shouldn't that happen on the third case here?
Why does a hard fail get full credit?

You may want to consider combining these two tests into one and adding a parameter that controls the return of the submission.